client/asset/eth/eth.go

```diff
@@ -338,15 +337,92 @@ func (eth *ExchangeWallet) OwnsAddress(address string) (bool, error) {
 
 // Balance returns the total available funds in the account. The eth node
 // returns balances in wei. Those are flored and stored as gwei, or 1e9 wei.
-//
-// TODO: Return Immature and Locked values.
 func (eth *ExchangeWallet) Balance() (*asset.Balance, error) {
 	eth.lockedFundsMtx.Lock()
 	defer eth.lockedFundsMtx.Unlock()
 
 	return eth.balanceImpl()
 }
 
+// txEffectOnBalance calculates the effects of a transaction on the balance
+// of the sender. It only takes into account the transaction types expected
+// to be sent by this wallet, including withdrawals, initiations, redeems,
+// and refunds.
+func (eth *ExchangeWallet) txEffectOnBalance(tx *types.Transaction) (incoming *big.Int, outgoing *big.Int, err error) {
+	incoming, outgoing = new(big.Int), new(big.Int)
+
+	// Max possible gas fee
+	feeCap := tx.GasFeeCap()
+	gas := big.NewInt(int64(tx.Gas()))
+	outgoing.Mul(feeCap, gas)
+
+	// If the tx contains value, it is either an initiation or withdrawal.
+	// In either case we don't need to parse the tx data, so we can return
+	// here.
```

Comment:
Won't any transaction that transfers value from the sender have a non-zero `tx.Value()`? Will this catch all outgoing, regardless of type?

Reply:
I think that's the interpretation of "withdrawal" here.

What I'm struggling to understand is how do we know we are the sender in this `types.Transaction`.  Its not yet clear to me that `eth_pendingTransactions` only pertains to this account's relevant transactions.  Do we need a `Signer` to determine the sender?

Reply:
Ultimately, we'll be using `(*light.TxPool).GetPoolTransactions`. The docs for `light.TxPool` say 

```
// TxPool implements the transaction pool for light clients, which keeps track
// of the status of locally created transactions...
```

Maybe we should check the `From` address, though.

Reply:
What about deposits then?  Those aren't locally-created transactions.

Reply:
That's true. Deposits will need a confirmation to be seen by the wallet, I guess.

Reply:
Apparently so: https://github.com/ethereum/go-ethereum/issues/16216#issuecomment-369191599

Reply:
BTW, in `startNode`, why don't we need to track and stop the light client:

```go
light, err := les.New(node, &ethCfg)
	if err != nil {
		return err
	}
	// on shutdown, light.BlockChain().Stop
```

Reply:
Nevermind that q about the light client.  I see it is registered with `Node`